File: national_id.js
function isValidIranianNationalCode(input) {
  if (!/^\d{10}$/.test(input)) return false;
  const check = +input[9];
  const sum = input.split('').slice(0, 9).reduce((acc, x, i) => acc + +x * (10 - i), 0) % 11;
  return sum < 2 ? check === sum : check + sum === 11;
}

function isValidIranianNationalLegalCode(input) {
    if (!/^\d{11}$/.test(input)) return false;
    const check = +input[10];
    const decimal = +input[9]+2;
    const coefficient = [29,27,23,19,17,29,27,23,19,17]
    let sum = input.split('').slice(0, 10).reduce((acc, x, i) => acc + ((decimal + +x) * coefficient[i]), 0) % 11;
    if(sum==10)
        sum = 0;
    return sum === check;
}

File: national_id.py
import re

def valid_n_id(num):
    try: num = f"{int(num):010d}"
    except: return False
    if not re.search(r'^\d{10}$', num):
        return False
    
    check = int(num[-1])
    s = sum([int(num[x]) * (10 - x) for x in range(9)]) % 11
    return (s < 2 and check == s) or (s >= 2 and check + s == 11)


def valid_legal_n_id(num):
    try: num = f"{int(num):011d}"
    except: return False
    if not re.search(r'^\d{11}$', num):
        return False
    check = int(num[-1])
    decimal = int(num[-2]) + 2
    coefficient = [29,27,23,19,17,29,27,23,19,17]
    s = sum([(int(num[x]) + decimal) * coefficient[x] for x in range(10)]) % 11
    if s == 10: s = 0
    return s == check

